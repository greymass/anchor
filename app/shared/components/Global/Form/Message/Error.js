// @flow
import React, { Component } from 'react';
import { translate } from 'react-i18next';
import { Button, Message, Header } from 'semantic-ui-react';

class FormMessageError extends Component<Props> {
  render() {
    const {
      error,
      errors,
      onClose,
      t
    } = this.props;

    const errorMessages = (errors || []);

    errorMessages.push(error);

    return (error || errors)
      ? (
        <div>
          <Message negative>
            <Header>{t('error')}</Header>
            { errorMessages.map((err) => <p key={err}>{t(`error_${err}`)}</p>) }
          </Message>

          {(onClose) ? (
            <Button
              color="red"
              content={t('close')}
              fluid
              onClick={onClose}
            />) : ''}
        </div>
      ) : '';
  }
}

export default translate('global')(FormMessageError);
