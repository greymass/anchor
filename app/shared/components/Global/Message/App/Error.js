// @flow
import React, { Component } from 'react';

import { withTranslation } from 'react-i18next';

import { Message, Header, Segment } from 'semantic-ui-react';

class FormMessageError extends Component<Props> {
  render() {
    const {
      error,
      t,
    } = this.props;

    return (
      <Segment>
        <Header>{t('global_message_error_header')}</Header>
        {typeof error === 'object'
          ? (
            <Message negative>
              <Header>{error.message}</Header>
              <p>{error.stack}</p>
            </Message>
          )
          : (
            <Message negative>
              <Header>
                {error}
              </Header>
            </Message>
          )
        }
      </Segment>
    );
  }
}

export default withTranslation('global')(FormMessageError);
