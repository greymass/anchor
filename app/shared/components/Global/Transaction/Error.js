// @flow
import React, { Component } from 'react';
import { translate } from 'react-i18next';

import GlobalTransactionErrorAuthorization from './Error/Authorization';
import GlobalTransactionErrorDefault from './Error/Default';

const transactionErrorsMapping = {
  authorization: GlobalTransactionErrorAuthorization
};

class GlobalTransactionError extends Component<Props> {
  constructor(props) {
    super(props);
    const {
      error
    } = this.props;

    this.state = {
      ComponentType: transactionErrorsMapping[error] || GlobalTransactionErrorDefault
    };
  }

  render() {
    const {
      error,
      onClose,
      style
    } = this.props;

    return (
      <ComponentType
        error={error}
        onClose={onClose}
        style={style}
      />
    );
  }
}

export default GlobalTransactionError;
