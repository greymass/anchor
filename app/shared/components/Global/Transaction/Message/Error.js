// @flow
import React, { Component } from 'react';
import { withTranslation } from 'react-i18next';
import { connect } from 'react-redux';
import { Button } from 'semantic-ui-react';
import compose from 'lodash/fp/compose';

import GlobalTransactionErrorAuthorization from './Error/Authorization';
import GlobalTransactionErrorDefault from './Error/Default';
import GlobalTransactionErrorCpuUsage from './Error/CpuUsage';
import GlobalTransactionErrorNetUsage from './Error/NetUsage';
import GlobalTransactionErrorLedgerBusy from './Error/Ledger/Busy';
import GlobalTransactionErrorLedgerData from './Error/Ledger/Data';
import GlobalTransactionErrorLedgerError from '../../../../containers/Global/Hardware/Ledger/Error';
import GlobalTransactionErrorResourceUsage from './Error/ResourceUsage';

const transactionErrorsMapping = {
  unsatisfied_authorization: GlobalTransactionErrorAuthorization,
  leeway_deadline_exception: GlobalTransactionErrorCpuUsage,
  tx_net_usage_exceeded: GlobalTransactionErrorNetUsage,
  tx_cpu_usage_exceeded: GlobalTransactionErrorCpuUsage,
  resource_usage: GlobalTransactionErrorResourceUsage,
};

const transactionErrorMsgMapping = {
  'transport.decorateAppAPIMethods is not a function': GlobalTransactionErrorLedgerError,
  'Transport race condition': GlobalTransactionErrorLedgerBusy,
  'Ledger device: Invalid data received (0x6a80)': GlobalTransactionErrorLedgerData,
};

export class GlobalTransactionMessageError extends Component<Props> {
  constructor(props) {
    super(props);
    const {
      error,
      fuel,
      settings,
    } = this.props;

    let errorName = error.error && error.error.name;
    if (!errorName) {
      errorName = error.json && error.json.error && error.json.error.name;
    }

    // Handle case where it is greymassfuel error
    const isFuelError = error.error &&
      error.error.details &&
      JSON.stringify(error.error.details).includes('greymassfuel');

    if (isFuelError) {
      errorName = 'leeway_deadline_exception';
    }

    let component = GlobalTransactionErrorDefault;
    if (transactionErrorsMapping[errorName]) {
      component = transactionErrorsMapping[errorName];
    } else if (transactionErrorMsgMapping[error.message]) {
      component = transactionErrorMsgMapping[error.message];
    }

    const canProceedWithFee = (fuel && fuel.alternative && fuel.alternative.request && fuel.alternative.request.code === 402);
    if (canProceedWithFee) {
      component = transactionErrorsMapping.resource_usage;
    }

    this.state = {
      canProceedWithFee,
      ComponentType: component
    };
  }

  render() {
    const {
      error,
      onClose,
      style,
      t
    } = this.props;

    const {
      canProceedWithFee,
      ComponentType
    } = this.state;

    return (
      <div style={style}>
        <ComponentType
          error={error}
          onClose={onClose}
          style={style}
        />
        {(onClose && !canProceedWithFee) ? (
          <Button
            color="red"
            content={t('close')}
            fluid
            onClick={onClose}
          />) : ''}
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  fuel: state.fuel,
  settings: state.settings,
});

export default compose(
  withTranslation('global'),
  connect(mapStateToProps)
)(GlobalTransactionMessageError);
