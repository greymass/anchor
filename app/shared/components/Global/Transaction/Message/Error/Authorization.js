import React, { Component } from 'react';
import { translate } from 'react-i18next';
import { List } from 'semantic-ui-react';
import ReactJson from 'react-json-view';

class GlobalTransactionMessageErrorAuthorization extends Component<Props> {
  render() {
    const {
      error,
      t
    } = this.props;

    const colorStyle = { color: '#3C3A3B' };

    return (
      <div>
        <p>
          {t('global_transaction_message_error_authorization_paragraph')}
        </p>
        <hr style={colorStyle} />
        <List style={colorStyle} ordered>
          <List.Item>
            {t('global_transaction_message_error_authorization_list_item_one')}
          </List.Item>
          <List.Item>
            {t('global_transaction_message_error_authorization_list_item_two')}
          </List.Item>
          <List.Item>
            {t('global_transaction_message_error_authorization_list_item_three')}
          </List.Item>
        </List>
        <hr style={colorStyle} />
        <p style={colorStyle}>*&nbsp;{t('global_transaction_message_error_authorization_note')}</p>
        {(typeof error === 'object') ? (
          <ReactJson
            collapsed={2}
            displayDataTypes={false}
            displayObjectSize={false}
            iconStyle="square"
            name={null}
            src={error}
            style={{ padding: '1em' }}
          />
        ) : ''}
      </div>
    );
  }
}

export default translate('global')(GlobalTransactionMessageErrorAuthorization);
