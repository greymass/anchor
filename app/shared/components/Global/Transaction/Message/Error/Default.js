import React, { Component } from 'react';
import { withTranslation } from 'react-i18next';
import { Header, Message } from 'semantic-ui-react';
import ReactJson from 'react-json-view';

export class GlobalTransactionMessageErrorDefault extends Component<Props> {
  render() {
    const {
      error,
      t
    } = this.props;

    let errorMessage = error.error && error.error.details && error.error.details[0] && error.error.details[0].message;

    if (errorMessage) {
      const errorMessageArray = errorMessage.split('assertion failure with message:');
      errorMessage = errorMessageArray[1] &&
        errorMessageArray[1]
          .trim()
          .split(' ')
          .slice(0, 8)
          .join('_');
      // The above line will basically take the first 8 words
      // of message and turn it into a locale key.
      errorMessage = errorMessage && errorMessage.replace(/\(|\)|%/g, '');
      errorMessage = errorMessage || errorMessageArray[0];
    }
    return (
      <Message negative>
        <Header>{t('error')}</Header>
        <p key={error}>
          {(errorMessage)
            ? t(errorMessage)
            : `${error.message} (${error.error ? error.error : ''})`
          }
        </p>
        {(typeof error === 'object' && !(error instanceof Error)) ? (
          <ReactJson
            collapsed={1}
            collapseStringsAfterLength={60}
            displayDataTypes={false}
            displayObjectSize={false}
            iconStyle="square"
            name={null}
            src={error}
            style={{ padding: '1em' }}
          />
        ) : ''}
      </Message>
    );
  }
}

export default withTranslation('global')(GlobalTransactionMessageErrorDefault);
