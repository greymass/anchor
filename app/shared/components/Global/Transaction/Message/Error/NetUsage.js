import React, { Component } from 'react';
import { withTranslation } from 'react-i18next';
import { Header, Icon, List, Segment } from 'semantic-ui-react';
import ReactJson from 'react-json-view';

export class GlobalTransactionMessageErrorNetUsage extends Component<Props> {
  render() {
    const {
      error,
      t
    } = this.props;
    return (
      <Segment basic color="red" size="large">
        <Header size="medium">
          <Icon name="warning sign" />
          <Header.Content>
            Insufficient Network Resources (NET)
            <Header.Subheader>
              This account currently lacks the required network resources to perform this transaction.
            </Header.Subheader>
          </Header.Content>
        </Header>
        <p>Address this issue by performing one of the following, then try again:</p>
        <List relaxed bulleted>
          <List.Item>Rent more NET using the Resources interface and try again.</List.Item>
          <List.Item>Wait for your existing NET to replenish within the next 24 hours.</List.Item>
        </List>
        {(typeof error === 'object' && error.json) ? (
          <ReactJson
            collapsed={1}
            displayDataTypes={false}
            displayObjectSize={false}
            iconStyle="square"
            name={null}
            src={error.json.error}
            style={{ padding: '1em' }}
          />
        ) : ''}
      </Segment>
    );
  }
}

export default withTranslation('global')(GlobalTransactionMessageErrorNetUsage);
