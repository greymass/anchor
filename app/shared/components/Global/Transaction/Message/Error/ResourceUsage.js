// @flow
import { get } from 'dot-prop-immutable';
import React, { PureComponent } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { withTranslation } from 'react-i18next';
import compose from 'lodash/fp/compose';
import { Button, Checkbox, Icon, Header, List, Segment } from 'semantic-ui-react';
import { isEmpty, sum, sumBy } from 'lodash';

import { retryWithFee } from '../../../../../actions/fuel';
import { setSetting } from '../../../../../actions/settings';

class GlobalTransactionMessageErrorResourceUsage extends PureComponent<Props> {
  componentDidMount = () => {
    const { fuel, settings, system } = this.props;
    if (fuel.alternative && settings.transactionFees) {
      this.props.actions.retryWithFee(system.latestFailure, fuel.alternative.request);
    }
  }
  componentDidUpdate(prevProps) {
    const { fuel, settings, system } = this.props;
    if (
      !prevProps.fuel.alternative
      && fuel.alternative
    ) {
      if (settings.transactionFees) {
        this.props.actions.retryWithFee(system.latestFailure, fuel.alternative.request);
      }
    }
  }
  toggle = (e, { checked }) => {
    this.props.actions.setSetting('transactionFees', checked);
  }
  prompt = () => {
    const { fuel, system } = this.props;
    this.props.actions.retryWithFee(system.latestFailure, fuel.alternative.request);
  }
  render() {
    const {
      fuel,
      system,
      t
    } = this.props;
    const canProceedWithFee = (fuel && fuel.alternative && fuel.alternative.request && fuel.alternative.request.code === 402);
    const loading = !fuel || !fuel.alternative;
    return (
      <Segment basic loading={loading}>
        <Header
          content="Transaction Fee"
          icon="stop"
          size="large"
          subheader="Your account lacks the network resources for this transaction to succeed."
        />
        {(loading)
          ? (
            <p>Loading...</p>
          )
          : (
            <Segment stacked size="large" textAlign="center">
              <p>To perform this transaction, you will either need to pay a fee or use the Resources section of the wallet to manage your account resources.</p>
              <Header
                content={fuel.alternative.request.data.fee}
                subheader="Transaction Fee"
              />
              <Button
                content={`Proceed with ${fuel.alternative.request.data.fee} fee`}
                onClick={this.prompt}
                primary
                size="large"
              />
              <div style={{ margin: '1em 0' }}>
                <Checkbox
                  label="Automatically display fee-based transactions."
                  onChange={this.toggle}
                  size="large"
                />
              </div>
            </Segment>
          )
        }
      </Segment>
    );
  }
}

const mapStateToProps = (state) => ({
  fuel: state.fuel,
  settings: state.settings,
  system: state.system,
});

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      retryWithFee,
      setSetting,
    }, dispatch)
  };
}

export default compose(
  withTranslation('global'),
  connect(mapStateToProps, mapDispatchToProps)
)(GlobalTransactionMessageErrorResourceUsage);
