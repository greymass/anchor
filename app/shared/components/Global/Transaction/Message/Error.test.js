import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { GlobalTransactionMessageError } from './Error';

const mockProps = {
  error: 'error message',
  onClose: () => null,
  style: { margin: '10px' },
  t: text => text,
  transaction: {},
};

describe('GlobalTransactionMessageError', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<GlobalTransactionMessageError {...mockProps} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

