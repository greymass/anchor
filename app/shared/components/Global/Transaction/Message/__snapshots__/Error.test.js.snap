// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GlobalTransactionMessageError renders correctly 1`] = `
<div
  style={
    Object {
      "margin": "10px",
    }
  }
>
  <Message
    negative={true}
  >
    <Header>
      error
    </Header>
    <Translate(GlobalTransactionMessageErrorDefault)
      error="error message"
      onClose={[Function]}
      style={
        Object {
          "margin": "10px",
        }
      }
    />
  </Message>
  <Button
    as="button"
    color="red"
    content="close"
    fluid={true}
    onClick={[Function]}
  />
</div>
`;
